Devel/rbtree

Werkov opened the pull request.

jeffmahoney committed:
types: Add basic rbtree traversal

Add functions for traversing generic RB tree structures and accompanying
tests (both for correctness and applicability to real kernel dumps).

Signed-off-by: Jeff Mahoney <jeffm@suse.com>
Signed-off-by: Michal Koutný <mkoutny@suse.com>

Werkov committed:
git: Ignore vim swap files

Signed-off-by: Michal Koutný <mkoutny@suse.com>

jeffmahoney commented on crash/types/rbtree.py:
```diff
@@ -0,0 +1,82 @@
+    if root.type == types.rb_root_type.pointer():
+        root = root.dereference()
+    elif root.type != types.rb_root_type:
+        raise ArgumentTypeError('root', root, types.rb_root_type)
```
This should be UnexpectedGDBTypeError now.

jeffmahoney commented on crash/types/rbtree.py:
```diff
@@ -0,0 +1,82 @@
+
+def rbtree_postorder_for_each(root: gdb.Value) -> Iterable[gdb.Value]:
+    if not isinstance(root, gdb.Value):
+        raise InvalidArgumentError("root must be gdb.Value representing 'struct rb_root' or 'struct rb_root *' not {}"
```
This should be ArgumentTypeError now

jeffmahoney reviewed (changes requested):
There are a few nits, comments inline, but also (as you mentioned on IRC) since this is intended to be part of the API, we need documentation for it.  We also need test cases. If you can come up with simple cases like those in tests/test_list.py, that would be helpful.  Test cases that work with common in-kernel rbtrees would also be helpful to ensure that it still works correctly when the kernel is updated.

Finally, outside of the .gitignore changes, this should all be a single commit.

Thanks,

-Jeff

Werkov commented:
No ready yet (only backing up the branch), will rebase later.

Werkov commented:
Force pushed a new version:
- rebased onto current master,
- added unit and dump tests,
- squashed altogether.

Also, I noticed that similar API for linked list traversing says in the docs:

> def list_for_each
>         [...]
>         gdb.Value: The next node in the list.  The value is
>         of type ``struct list_head``.

However, the returned gdb.Value is of type `struct list_head *`, i.e. the docs isn't consistent with the implementation. The rbtree implementation honors the docs. Let me know which is the preferred way for these iterators (I expect the non-pointer variant).

Werkov committed:
crash.types.list: Unify return type of iterator

We honor the principle of accepting both pointer to or object itself on
input and giving out the object on output. The list iterator
implementation returns the pointer (contradicting docs). Switch to the
object type for consistency with other APIs (e.g. list_for_each_entry).

Signed-off-by: Michal Koutný <mkoutny@suse.com>

Werkov commented:
I've rebased the branch on the current master (resolved trouble with gdb update from 8.3 to 9.1) and added one more commit to rectify the type situation as laid out in my previous comment.
AFAICT, all the remarks should be resolved now.
